enhancements/kube-apiserver/enforce-labels.md

```diff
@@ -0,0 +1,198 @@
+---
+title: enforce-labels
+authors:
+ - "@mmirecki"
+reviewers:
+ - "@fedepaol"
+approvers:
+ - TBD
+creation-date: 2020-08-07
+last-updated: 2020-08-07
+status: implementable
+---
```

Comment:
I miss a discussion about risk of this central component in the critical path of many resources (availability / HA concept):
- how does this avoid to put the control-plane components at risk?
- how does this avoid rendering the cluster unavailable for an extended amount of time (minutes?) when the node with the webhook is hard-killed?
- what is the influence on performance / scale?

Reply:
**how does this avoid to put the control-plane components at risk?**

In both solutions (gatekeeper and in-house) the webhook would avoid touching the control-plane by excluding it from the webhook with the namespace-selector.

Gatekeeper.
Gatekeeper uses a single admission webhook to for constraints enforcement. I assume the same solution will be used for the mutation feature. The control-plane resources can be excluded from gatekeeper interference by specifying an appropriate namespace selector on the webhook. 
Gatekeeper would be installed/configured by an operator, which would adjust the webhook namespace selector (based on a GatekeeperConfig CRD?). 
The default gatekeeper webhook configuration is using "failurePolicy​: ​Ignore", I assume this would be extended to the mutation feature (more on this: https://github.com/open-policy-agent/gatekeeper#admission-webhook-fail-open-status). Any resources which would slip into the system unmodified would later be corrected using the additional relabeling feature (implemented in the second phase).

In-house solution
With the in-house solution we would be creating the mutating webhooks dynamicly, one per namespace (maybe even per resource). The webhooks would therefore only affect the resources they are configured to modify. With an empty configuration, there would be no webhooks at all.

**how does this avoid rendering the cluster unavailable for an extended amount of time (minutes?) when the node with the webhook is hard-killed?**

Gatekeeper
With "failurePolicy​: ​Ignore" we would allow unmodified resources to be created. In the second phase, these would be corrected by the relabeling functionality.

In-house solution
Only some of the namespace/resource webhooks would be killed. We could also make use of "failurePolicy​: ​Ignore" to allow unmodified resources to be added.

**what is the influence on performance / scale?**

Gatekeeper
Work is being done on multi-pod deployments of Gatekeeper. I need to inquire on the eta of this.

Reply:
One additional note: if we drive the webhook's namespaces selector via an operator, we can (and should) have `failurePolicy: Fail` as we will be sure it's not going to impact control plane components.

Reply:
Thanks for the answers. Would copy those into the doc.

Reply:
Done